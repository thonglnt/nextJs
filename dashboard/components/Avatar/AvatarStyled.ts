import { Badge, styled } from '@mui/material'

interface BadgeStyledProps {
  status: 'on' | 'off'
}

export const StyledBadge = styled(Badge)<BadgeStyledProps>`
  ${({ status, theme }) => ({
    '& .MuiBadge-badge': {
      backgroundColor: `${
        status === 'on' ? '#44b700' : `${theme.palette.grey[400]}`
      }`,
      color: `${status === 'on' ? '#44b700' : `${theme.palette.grey[400]}`}`,
      boxShadow: `0 0 0 2px ${theme.palette.background.paper}`,
      '&::after': {
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        borderRadius: '50%',
        border: '1px solid currentColor',
        content: '""',
      },
    },
    '@keyframes ripple': {
      '0%': {
        transform: 'scale(.8)',
        opacity: 1,
      },
      '100%': {
        transform: 'scale(2.4)',
        opacity: 0,
      },
    },
  })}
`
