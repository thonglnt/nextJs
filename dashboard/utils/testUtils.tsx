import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'

// Themes
import theme from '@self-types/Themes.types'

export const customRender = (children: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
}

export * from '@testing-library/react'
export { customRender as render }
