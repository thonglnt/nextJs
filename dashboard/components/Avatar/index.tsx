'use client'

import { Avatar, AvatarProps } from '@mui/material'
import { memo } from 'react'

// Utils
import { stringAvatar } from '@utils/index'

// Styled
import { StyledBadge } from './AvatarStyled'

export interface IAvatar extends AvatarProps {
  type: 'normal' | 'badge' | 'letter'
  status?: 'on' | 'off'
}

const Avatars = ({ type, alt, src, status = 'off', ...props }: IAvatar) => {
  return (
    <>
      {type === 'badge' ? (
        <StyledBadge
          status={status}
          overlap="circular"
          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
          variant="dot"
        >
          <Avatar alt={alt} src={src} {...props} />
        </StyledBadge>
      ) : type === 'letter' ? (
        <Avatar {...stringAvatar('Kent Dodds')} {...props} />
      ) : (
        <Avatar alt={alt} src={src} {...props} />
      )}
    </>
  )
}

export default memo(Avatars)
