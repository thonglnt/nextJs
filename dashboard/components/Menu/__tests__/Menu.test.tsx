import { PersonAdd } from '@mui/icons-material'
import { MenuItem, Divider, ListItemIcon, Box } from '@mui/material'

// Utils
import { render, screen, fireEvent } from '@utils/testUtils'

// Components
import Avatar from '@components/Avatar'
import MenuComposition, { IMenuComposition } from '..'

describe('Menu component', () => {
  const mockProps: IMenuComposition = {
    horizontal: 'center',
    element: <Avatar type="letter">M</Avatar>,
    open: false,
    children: (
      <Box>
        <MenuItem>
          <Avatar type="normal" /> Profile
        </MenuItem>
        <MenuItem>
          <Avatar type="normal" /> My account
        </MenuItem>
        <Divider />
        <MenuItem>
          <ListItemIcon>
            <PersonAdd fontSize="small" />
          </ListItemIcon>
          Add another account
        </MenuItem>
      </Box>
    ),
  }

  it('should render Menu component', () => {
    const { container } = render(<MenuComposition {...mockProps} />)

    expect(container).toMatchSnapshot()
  })

  it('Should call onClick on Menu and open Menu', () => {
    render(<MenuComposition {...mockProps} />)

    const button = screen.getByRole('button')
    fireEvent.click(button)

    expect(screen.getByTestId('menu')).toBeInTheDocument()
  })

  it('Should call onClick on Menu and close Menu', () => {
    const { container } = render(<MenuComposition {...mockProps} />)

    const button = screen.getByRole('button')
    fireEvent.click(button)
    const menu = screen.getByTestId('menu')
    fireEvent.click(menu)

    expect(container).not.toHaveTextContent('Profile')
  })
})
