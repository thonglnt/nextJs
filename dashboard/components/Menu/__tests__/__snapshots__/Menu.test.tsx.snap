// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component should render Menu component 1`] = `
<div>
  <button
    aria-haspopup="true"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <div
      class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1vfbbup-MuiAvatar-root"
    >
      M
    </div>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
